### PR TITLE
MODSOURMAN-203: Map Mode of issuance

### PR DESCRIPTION
**Pull request by VRohach**
1. New mechanism for mapping "mode of issuance" was created.
2. Mode of issuance entity was created.

### CODE REVIEW

**Comment by VRohach** [comment]
Tests for this implementation currently in progress

File: mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java [hunk 1]
```diff
@@ -351,6 +353,52 @@ public String apply(RuleExecutionContext context) {
...
         .orElse(STUB_FIELD_TYPE_ID);
     }
+  },
+
+  SET_ISSUANCE_MODE_ID() {
+
+    @Override
+    public String apply(RuleExecutionContext context) {
+      String subFieldValue = context.getSubFieldValue();
+      char seventhChar = subFieldValue.charAt(6);
```
Comment (OleksiiKuzminov) [inline mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java]: comment

File: mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java [hunk 2]
```diff
@@ -351,6 +353,52 @@ public String apply(RuleExecutionContext context) {
...
+    private String findIssuanceModeId(List<IssuanceMode> issuanceModes, IssuanceModeEnum issuanceModeType,
+                                      String defaultId) {
+      return issuanceModes.stream()
+        .filter(issuanceMode -> issuanceMode.getName().equals(issuanceModeType.getValue()))
+        .findFirst()
+        .map(IssuanceMode::getId)
+        .orElse(defaultId);
+    }
+
+    private String matchIssuanceModeIdViaLeaderSymbol(char seventhChar, List<IssuanceMode> issuanceModes, String defaultId) {
```
Comment (OleksiiKuzminov) [inline mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java]: move to enum

File: mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java [hunk 3]
```diff
@@ -351,6 +353,52 @@ public String apply(RuleExecutionContext context) {
...
+        return StringUtils.EMPTY;
+      }
+      String defaultIssuanceModeId = findIssuanceModeId(issuanceModes, IssuanceModeEnum.UNSPECIFIED, StringUtils.EMPTY);
+      return matchIssuanceModeIdViaLeaderSymbol(seventhChar, issuanceModes, defaultIssuanceModeId);
+    }
+
+    private String findIssuanceModeId(List<IssuanceMode> issuanceModes, IssuanceModeEnum issuanceModeType,
+                                      String defaultId) {
+      return issuanceModes.stream()
+        .filter(issuanceMode -> issuanceMode.getName().equals(issuanceModeType.getValue()))
```
Comment (OleksiiKuzminov) [inline mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java]: equalsIgoneCase

File: mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java [hunk 4]
```diff
@@ -351,6 +353,52 @@ public String apply(RuleExecutionContext context) {
...
+
+  SET_ISSUANCE_MODE_ID() {
+
+    @Override
+    public String apply(RuleExecutionContext context) {
+      String subFieldValue = context.getSubFieldValue();
+      char seventhChar = subFieldValue.charAt(6);
+      List<IssuanceMode> issuanceModes = context.getMappingParameters().getIssuanceModes();
+      if (issuanceModes == null || issuanceModes.isEmpty()) {
+        return StringUtils.EMPTY;
```
Comment (OleksiiKuzminov) [inline mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java]: log debug cases with StringUtils.EMPTY

**Comment by OleksiiKuzminov** [comment]
unit tests

File: mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java [hunk 5]
```diff
@@ -351,6 +353,52 @@ public String apply(RuleExecutionContext context) {
...
+  },
+
+  SET_ISSUANCE_MODE_ID() {
+
+    @Override
+    public String apply(RuleExecutionContext context) {
+      String subFieldValue = context.getSubFieldValue();
+      char seventhChar = subFieldValue.charAt(6);
+      List<IssuanceMode> issuanceModes = context.getMappingParameters().getIssuanceModes();
+      if (issuanceModes == null || issuanceModes.isEmpty()) {
```
Comment (KaterynaSenchenko) [inline mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java]:
```suggestion
      if (CollectionUtils.isEmpty(issuanceModes)) {
```

Comment (VRohach) [reply mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java]: I implement this check like in other Normalization-functions for common approach. But I think that `CollectionUtils.isEmpty()` -  is the best way to do it. So, I will change it in other normalization-function for saving a common approach. Thx @KaterynaSenchenko !

File: mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java (hunk 4 above)
Comment (VRohach) [reply mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java]: Yes, sure. I added debug logging for each case where we will have an empty string as a result of normalization function.

File: mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java (hunk 3 above)
Comment (VRohach) [reply mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java]: Done

File: mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java (hunk 2 above)
Comment (VRohach) [reply mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java]: Done. @OleksiiKuzminov could you please review my approach?

File: mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java (hunk 1 above)
Comment (VRohach) [reply mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java]: Added.

File: mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java (hunk 5 above)
Comment (VRohach) [reply mod-source-record-manager-server/src/main/java/org/folio/services/mappers/processor/functions/NormalizationFunction.java]: I double-checked and found that it was a really specific case. Thank you for this comment!

**Pull request merged**